### PR TITLE
Blazor .NET to JS Streaming Interop

### LINKED ISSUE #23039: Blazor .NET to JS Streaming Interop
https://github.com/dotnet/aspnetcore/pull/34817

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 2964d558-cd6d-7421-f7b9-cb939792c2cf
* Version Independent ID: a3517e56-69fc-1ec7-b55a-dbb54d68a807
* Content: [Call .NET methods from JavaScript functions in ASP.NET Core Blazor](https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/javascript-interoperability/call-dotnet-from-javascript.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #23039
Addresses #22045 

I also made some improvements to the earlier text/example that we added for streaming to .NET.

btw -- need review of https://github.com/dotnet/AspNetCore.Docs/pull/23033.

### CODE REVIEW

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md [hunk 1]
```diff
@@ -657,34 +657,79 @@ Other data types, such as string arrays, can be converted but require creating a
 
 ## Stream from JavaScript to .NET
 
-Blazor supports streaming data directly from JavaScript to .NET. Streams are requested using the `IJSStreamReference` interface:
+Blazor supports streaming data directly from JavaScript to .NET. Streams are requested using the `Microsoft.JSInterop.IJSStreamReference` interface.
 
-In client-side JavaScript:
+`Microsoft.JSInterop.IJSStreamReference` returns a <xref:System.IO.Stream> and uses the following parameters:
```
Comment (TanayParikh) [inline aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:662]: Not sure of doc convention here, but maybe:

```suggestion
`Microsoft.JSInterop.IJSStreamReference.OpenReadStreamAsync` returns a <xref:System.IO.Stream> and uses the following parameters:
```

https://github.com/dotnet/aspnetcore/blob/66ceea4fc6ccbd710c5b77782761f92557a08be6/src/JSInterop/Microsoft.JSInterop/src/IJSStreamReference.cs#L27

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md [hunk 2]
```diff
@@ -657,34 +657,79 @@ Other data types, such as string arrays, can be converted but require creating a
...
 await dataReferenceStream.CopyToAsync(outputFileStream);
 ```
 
-`IJSStreamReference` returns a <xref:System.IO.Stream> with the following parameters:
+In the preceding example:
 
-* `maxAllowedSize`: Maximum number of bytes permitted for the read from JavaScript. The value defaults to 512,000 bytes.
-* `cancellationToken`: A <xref:System.Threading.CancellationToken> for cancelling the read.
+* `JS` is an injected <xref:Microsoft.JSInterop.IJSRuntime> instance.
+* The JS stream is written to disk (`file.txt`) at the current user's temporary folder path (<xref:System.IO.Path.GetTempPath%2A>).
```
Comment (TanayParikh) [inline aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:691]:
```suggestion
* The `dataReferenceStream` is written to disk (`file.txt`) at the current user's temporary folder path (<xref:System.IO.Path.GetTempPath%2A>).
```

What's the %2A `*` at the end?

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md [hunk 3]
```diff
@@ -657,34 +657,79 @@ Other data types, such as string arrays, can be converted but require creating a
...
+
+## Stream from .NET to JavaScript
+
+*This feature applies to ASP.NET Core 6.0 Release Candidate 1 or later. ASP.NET Core 6.0 Release Candidate 1 is scheduled for release in September. ASP.NET Core 6.0 is scheduled for release later this year.*
+
+Blazor supports streaming data directly from .NET to JavaScript. Streams are created using a `Microsoft.JSInterop.DotNetStreamReference`.
+
+`Microsoft.JSInterop.DotNetStreamReference` represents a .NET stream and uses the following parameters:
+
+* `stream`: The stream sent to JavaScript.
```
Comment (TanayParikh) [inline aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:701]:
```suggestion
* `stream`: The stream to be sent JavaScript.
```

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md [hunk 4]
```diff
@@ -657,34 +657,79 @@ Other data types, such as string arrays, can be converted but require creating a
...
+In C# code:
+
+```csharp
+using var streamRef = new DotNetStreamReference(stream: {STREAM}, leaveOpen: false);
+await JS.InvokeVoidAsync("streamToJavaScript", streamRef);
+```
+
+In the preceding example:
+
+* The `{STREAM}` placeholder is a <xref:System.IO.Stream>.
```
Comment (TanayParikh) [inline aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:731]:
```suggestion
* The `{STREAM}` placeholder represents the <xref:System.IO.Stream> to be sent to JavaScript.
```

**Review by TanayParikh — APPROVED** [review]
👏 Thanks!

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md (hunk 2 above)
Comment (guardrex) [reply aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:691]: Convention for linking API docs. The doc has multiple members/overloads in it, so the asterisk represents any of them. It must be encoded to make linting happy, but it actually will work if we just use `*` ... you may see a few sprinkled around the doc set at large.

File: aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md (hunk 3 above)
Comment (guardrex) [reply aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:701]: We try to avoid state verbs if possible.

Comment (guardrex) [reply aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:701]: :smile: ... but it doesn't always work out.

> should be left open

🙈 ... I might change that.

Comment (TanayParikh) [reply aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:701]: Ah okay, feel free to ignore in that case 😄

Comment (guardrex) [reply aspnetcore/blazor/javascript-interoperability/call-javascript-from-dotnet.md:701]: I'll update tomorrow morning as I'm OOF the rest of the day. But the text is probably better more like .......

> When set to `true`, the stream is left open after transmission. Defaults to `false` if not provided.

**Pull request merged**